### PR TITLE
Cleanup leftovers for WAN replication SPI

### PR DESCRIPTION
**Pull request by mmedenjak**
- removed Node parameter from WanReplicationConsumer. Node is private
API and implementations may implement HazelcastInstanceAware to get
something like configuration.
- changed type of AbstractWanPublisherConfig.implementation to
WanReplicationPublisher and WanConsumerConfig.implementation to
WanReplicationConsumer.
- moved republishReplicationEvent, publishAntiEntropyEvent,
pause/stop/resume, getStats and removeWanEvents methods from public API.
These methods are all related to our WAN implementation (WAN event
republication, anti-entropy, lifecycle methods, statistics and clearing
the WAN data) and the user need not implement these methods to have the
simplest functioning WAN replication. Some of these are still exposed as
REST URIs (clear, pause/stop/resume, sync...) but the REST handlers will
be no-op in the case the user provides a custom publisher implementation.
It may be possible we expose these methods as public SPI in the future
so users may try to implement it and integrate with the REST handlers
and other subsystems.
- copied most migration and replication-related code from EE
WanReplicationService to OS. This now allows the user to fully implement
all methods on WAN publisher SPI, regardless if they are using OS or EE.
- moved all migration and replication-related methods to a
MigrationAwareWanReplicationPublisher interface. Now, if the user
wants the publisher to participate in migrations and replication, they
will need to implement this interface in addition to the
WanReplicationPublisher. Because of this, WanReplicationPublisher is
even simpler to implement.

Leftovers:
- code duplication between OS and EE migration and replication related
code. This can be addressed in a minor release.
- exposing event data as public methods on WAN replication events

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3363

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java [hunk 1]
```diff
@@ -268,4 +285,130 @@ public void removeWanEventCounters(String serviceName, String objectName) {
...
+
+        final Set<ServiceNamespace> namespaces = new HashSet<>();
+        for (DelegatingWanReplicationScheme publisher : wanReplications.values()) {
+            publisher.collectAllServiceNamespaces(event, namespaces);
+        }
+        return namespaces;
+    }
+
+    @Override
+    public boolean isKnownServiceNamespace(ServiceNamespace namespace) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java]: why is this for only map-service? Can you please add a short explanatory javaDoc on this method?

File: hazelcast/src/main/java/com/hazelcast/wan/impl/DelegatingWanReplicationScheme.java [hunk 2]
```diff
@@ -101,21 +102,31 @@ public void publishReplicationEventBackup(WanReplicationEvent event) {
     /**
      * Publishes a replication event to all publishers to which this publisher
      * delegates.
+     * Silently skips publishers not supporting republication.
      */
     public void republishReplicationEvent(WanReplicationEvent wanReplicationEvent) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/wan/impl/DelegatingWanReplicationScheme.java]: not used method?

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/DelegatingWanReplicationScheme.java]: also `addPublisher` is not used in this class?

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/DelegatingWanReplicationScheme.java]: + same for `destroyMapData`

File: hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java [hunk 3]
```diff
@@ -268,4 +285,130 @@ public void removeWanEventCounters(String serviceName, String objectName) {
         // EE WAN replication not supported in OS
         return Collections.emptyList();
     }
+
+    @Override
+    public void beforeMigration(PartitionMigrationEvent event) {
+        for (DelegatingWanReplicationScheme wanReplication : wanReplications.values()) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java]: seems we have streaming api candidates in this class?

File: hazelcast/src/main/java/com/hazelcast/wan/impl/DelegatingWanReplicationScheme.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/DelegatingWanReplicationScheme.java]: if these are overridden in EE, maybe we can put a short comment for that.

File: hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java]: On OS side, we only generate WAN events for map service. I'll add the javadoc.

File: hazelcast/src/main/java/com/hazelcast/wan/impl/DelegatingWanReplicationScheme.java (hunk 2 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/DelegatingWanReplicationScheme.java]: All of these are used from EE. I might separate them into an OS and EE `DelegatingWanReplicationScheme`/`EnterpriseDelegatingWanReplicationScheme` but with the previous WAN SPI, I've started not liking separate classes for OS and EE too much :)

I'll add the comment then.

File: hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java (hunk 3 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanReplicationServiceImpl.java]: True, it does seem repetitive.

File: hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventContainerReplicationOperation.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            service.appendWanReplicationConfig(wanConfig);
+        }
+
+        for (Entry<String, Map<String, Object>> wanReplicationSchemeEntry : eventContainers.entrySet()) {
+            String wanReplicationScheme = wanReplicationSchemeEntry.getKey();
+            Map<String, Object> eventContainersByPublisherId = wanReplicationSchemeEntry.getValue();
+            for (Entry<String, Object> publisherEventContainer : eventContainersByPublisherId.entrySet()) {
+                String publisherId = publisherEventContainer.getKey();
+                Object eventContainer = publisherEventContainer.getValue();
+                WanReplicationPublisher publisher = service.getPublisherOrFail(wanReplicationScheme, publisherId);
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventContainerReplicationOperation.java]: What if we fail during `getPublisherOrFail` call in the middle of for loop? In that case, we can have some publishers called method `processEventContainerReplicationData` and some others are skipped.
Is this problematic or is it better to check first if we can have a fail-able publisher and crash accordingly? So is it worth to separate logic in this class into 2 steps like this: validate publishers and process them?

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventContainerReplicationOperation.java]: We haven't encountered that issue before (I guess that would probably mean the configuration is different between members) but yes, we can do this to improve the situation.

**Comment by blazember** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventContainerReplicationOperation.java (hunk 4 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventContainerReplicationOperation.java]: Added the check, you can take a look at the last commit.

**Comment by mmedenjak** [comment]
Merging with [unrelated test failure](https://github.com/hazelcast/hazelcast/issues/16110). Thank you for the reviews, guys!

**Pull request merged**